presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java

```diff
@@ -368,8 +370,18 @@ private boolean isCandidate(Map<ColumnHandle, NullableValue> bindings)
             }
             LookupSymbolResolver inputs = new LookupSymbolResolver(assignments, bindings);
 
+            Object optimized;
+            try {
+                optimized = evaluator.optimize(inputs);
+            }
+            catch (PrestoException e) {
+                // Skip pruning if evaluation fails in a recoverable way.
+                // Failing here can cause spurious query failures for empty partitions.
+                TryFunction.propagateIfUnhandled(e);
```

Comment:
This looks as if only reusing  TryFunction's method for convenience, while what you seem to be doing is evaluating `try( filtering_condition )`. Maybe `TryFunction` should have a method
```
public static <T> T evaluate(Callable<T>)
```
that would make it explicit that you're reusing SQL `try` logic here.

Reply:
I wouldn't say convenience, but rather to have the exact same logic. Your suggestion is cleaner. I changed to `Supplier` to not have checked exceptions and needed to add a default value to return on failure:
```java
public static <T> T evaluate(Supplier<T> supplier, T defaultValue)